src/test/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSourceTest.java

```diff
@@ -1044,4 +1044,19 @@ public void testShouldRetrieveNullEvent() throws Exception {
     assertTrue(this.source.shouldRetrieve(mockSCMHeadObserver, null, PullRequestSCMHead.class));
     assertTrue(this.source.shouldRetrieve(mockSCMHeadObserver, null, BranchSCMHead.class));
   }
+
+  @Test
+  @Issue("JENKINS-67946")
+  public void testUserNamesWithAndWithoutUnderscores() {
+    // https://docs.github.com/en/enterprise-cloud@latest/admin/identity-and-access-management/managing-iam-for-your-enterprise/username-considerations-for-external-authentication#about-usernames-for-managed-user-accounts
+    // https://github.com/github/docs/blob/bfe96c289aee3113724495a2e498c21e2ec404e4/content/admin/identity-and-access-management/using-enterprise-managed-users-for-iam/about-enterprise-managed-users.md#about--data-variablesproductprodname_emus-
+    String repoOwnerWithUnderscore = "user_organization";
+    String repoOwnerNoUnderscore = "username";
+    String repoOwnerHyphen = "user-name";
+    String repoOwnerHyphenUnderscore = "user-name_organization";
+    assertTrue(repoOwnerWithUnderscore.matches(GitHubSCMSource.VALID_GITHUB_USER_NAME));
+    assertTrue(repoOwnerNoUnderscore.matches(GitHubSCMSource.VALID_GITHUB_USER_NAME));
+    assertTrue(repoOwnerHyphen.matches(GitHubSCMSource.VALID_GITHUB_USER_NAME));
+    assertTrue(repoOwnerHyphenUnderscore.matches(GitHubSCMSource.VALID_GITHUB_USER_NAME));
```

Comment:
I would actually add a lot more test example that validate the regexp against the username rules. Such as asserting that more than 39 characters fails, that double dashes or starting / ending with dashed don't work, etc...

For example:

```suggestion
    assertTrue("abcd".matches(GitHubSCMSource.VALID_GITHUB_USER_NAME));
    assertTrue("1234".matches(GitHubSCMSource.VALID_GITHUB_USER_NAME));
    assertTrue("user123".matches(GitHubSCMSource.VALID_GITHUB_USER_NAME));
    assertTrue("user123-org456".matches(GitHubSCMSource.VALID_GITHUB_USER_NAME));
    assertTrue("123-456".matches(GitHubSCMSource.VALID_GITHUB_USER_NAME));
    assertTrue("user123_org456".matches(GitHubSCMSource.VALID_GITHUB_USER_NAME));
    assertTrue("user123-org456-code789".matches(GitHubSCMSource.VALID_GITHUB_USER_NAME));
    assertTrue("user123-org456_code789".matches(GitHubSCMSource.VALID_GITHUB_USER_NAME));
    assertTrue("abcdefghijqlmnopkrstuvwxyz-123456789012".matches(GitHubSCMSource.VALID_GITHUB_USER_NAME));

    assertFalse("abcdefghijqlmnopkrstuvwxyz-1234567890123".matches(GitHubSCMSource.VALID_GITHUB_USER_NAME));
    assertFalse("user123@org456".matches(GitHubSCMSource.VALID_GITHUB_USER_NAME));
    assertFalse("user123.org456".matches(GitHubSCMSource.VALID_GITHUB_USER_NAME));
    assertFalse("user123--org456".matches(GitHubSCMSource.VALID_GITHUB_USER_NAME));
    assertFalse("user123-".matches(GitHubSCMSource.VALID_GITHUB_USER_NAME));
    assertFalse("-user123".matches(GitHubSCMSource.VALID_GITHUB_USER_NAME));
    assertFalse("user123__org456".matches(GitHubSCMSource.VALID_GITHUB_USER_NAME));
    assertFalse("user123_".matches(GitHubSCMSource.VALID_GITHUB_USER_NAME));
    assertFalse("_user123".matches(GitHubSCMSource.VALID_GITHUB_USER_NAME));
    assertFalse("user123-_org456".matches(GitHubSCMSource.VALID_GITHUB_USER_NAME));
    assertFalse("user123_org456-code789".matches(GitHubSCMSource.VALID_GITHUB_USER_NAME));
    assertFalse("user123_org456_code789".matches(GitHubSCMSource.VALID_GITHUB_USER_NAME));
```